cli/explain/compactgraph/compactgraph.go

```diff
@@ -0,0 +1,506 @@
+// used to compute the file digests, and an error if any.
+func ReadCompactLog(in io.Reader) (CompactGraph, string, error) {
+	d, err := zstd.NewReader(in)
+	if err != nil {
+		return nil, "", err
+	}
+	defer d.Close()
+	r := bufio.NewReader(d)
+
+	var hashFunction string
+	var entry spawnproto.ExecLogEntry
+	cg := make(CompactGraph)
+	previousInputs := make(map[int32]Input)
+	previousInputs[0] = emptyInputSet
+	for {
+		err = protodelim.UnmarshalFrom(r, &entry)
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, "", err
+		}
+		switch entry.Type.(type) {
+		case *spawnproto.ExecLogEntry_Invocation_:
+			hashFunction = entry.GetInvocation().HashFunctionName
+		case *spawnproto.ExecLogEntry_File_:
+			file := protoToFile(entry.GetFile(), hashFunction)
+			previousInputs[entry.Id] = file
+		case *spawnproto.ExecLogEntry_UnresolvedSymlink_:
+			symlink := protoToSymlink(entry.GetUnresolvedSymlink())
+			previousInputs[entry.Id] = symlink
+		case *spawnproto.ExecLogEntry_Directory_:
+			dir := protoToDirectory(entry.GetDirectory(), hashFunction)
+			previousInputs[entry.Id] = dir
+		case *spawnproto.ExecLogEntry_InputSet_:
+			inputSet := protoToInputSet(entry.GetInputSet(), previousInputs)
+			previousInputs[entry.Id] = inputSet
+		case *spawnproto.ExecLogEntry_Spawn_:
+			spawn, outputPaths := protoToSpawn(entry.GetSpawn(), previousInputs)
+			if spawn != nil {
+				for _, path := range outputPaths {
+					cg[path] = spawn
+				}
+			}
+		default:
+			panic(fmt.Sprintf("unexpected entry type: %T", entry.Type))
+		}
+	}
+	return cg, hashFunction, nil
+}
+
+func Diff(old, new CompactGraph) []*spawn_diff.SpawnDiff {
+	var spawnDiffs []*spawn_diff.SpawnDiff
+
+	oldPrimaryOutputs := old.primaryOutputs()
+	newPrimaryOutputs := new.primaryOutputs()
+
+	oldOnlyOutputs := old.reduceToRoots(setDifference(oldPrimaryOutputs, newPrimaryOutputs))
+	for _, path := range oldOnlyOutputs {
+		spawn := old[path]
+		spawnDiffs = append(spawnDiffs, &spawn_diff.SpawnDiff{
+			PrimaryOutput: spawn.PrimaryOutputPath(),
+			TargetLabel:   spawn.TargetLabel,
+			Mnemonic:      spawn.Mnemonic,
+			DiffType:      spawn_diff.SpawnDiff_OLD_ONLY,
+		})
+	}
+
+	newOnlyOutputs := new.reduceToRoots(setDifference(newPrimaryOutputs, oldPrimaryOutputs))
+	for _, path := range newOnlyOutputs {
+		spawn := new[path]
+		spawnDiffs = append(spawnDiffs, &spawn_diff.SpawnDiff{
+			PrimaryOutput: spawn.PrimaryOutputPath(),
+			TargetLabel:   spawn.TargetLabel,
+			Mnemonic:      spawn.Mnemonic,
+			DiffType:      spawn_diff.SpawnDiff_NEW_ONLY,
+		})
+	}
+
+	commonOutputs := setIntersection(oldPrimaryOutputs, newPrimaryOutputs)
+	type diffResult struct {
+		spawnDiff   *spawn_diff.SpawnDiff
+		localChange bool
+		affectedBy  []string
+	}
+	diffResults := sync.Map{}
+	diffWG := sync.WaitGroup{}
+	for _, output := range commonOutputs {
+		diffWG.Add(1)
+		go func() {
+			aSpawn := old[output]
```

Comment:
nit: where possible I like to put `defer wg.Done()` at the top, to minimize the distance between the `Add()` / `Done()` calls and prevent accidentally adding a `return` later that would prevent the `Done()` from getting called.

```suggestion
		diffWG.Add(1)
		go func() {
			defer diffWG.Done()	
			aSpawn := old[output]
```